aggregator/src/test/java/org/jenkinsci/plugins/workflow/BuildWatcher.java

```diff
@@ -0,0 +1,153 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow;
+
+import hudson.Extension;
+import hudson.console.AnnotatedLargeText;
+import hudson.console.LineTransformationOutputStream;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.junit.rules.ExternalResource;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+/**
+ * Echoes build output to standard error as it arrives.
+ * Usage: <pre>{@code @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();}</pre>
+ * Should work in combination with {@link JenkinsRule} or {@link RestartableJenkinsRule}.
+ */
+public final class BuildWatcher extends ExternalResource {
+
+    private static boolean active;
+    private static final Map<File,RunningBuild> builds = new ConcurrentHashMap<File,RunningBuild>();
+
+    private Thread thread;
+
+    @Override protected void before() throws Throwable {
+        active = true;
+        thread = new Thread("watching builds") {
+            @Override public void run() {
+                try {
+                    while (active) {
+                        for (RunningBuild build : builds.values()) {
+                            build.copy();
+                        }
+                        Thread.sleep(50);
+                    }
+                } catch (InterruptedException x) {
+                    // stopped
+                }
+                // last chance
+                for (RunningBuild build : builds.values()) {
+                    build.copy();
+                }
+            }
+        };
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    @Override protected void after() {
+        active = false;
+        thread.interrupt();
+    }
+
+    @Extension public static final class Listener extends RunListener<Run<?,?>> {
+
+        @Override public void onStarted(Run<?,?> r, TaskListener listener) {
+            RunningBuild build = new RunningBuild(r);
+            RunningBuild orig = builds.put(r.getLogFile(), build);
+            if (orig != null) {
+                System.err.println(r + " was started twice?!");
+            }
+        }
+
+        @Override public void onFinalized(Run<?,?> r) {
+            RunningBuild build = builds.remove(r.getLogFile());
+            if (build != null) {
+                if (active) {
```

Comment:
This is a general `@Extension`, so active even in tests that are not asking for me. In that case, `!active` and so the map is updated but nothing else happens. (Well, the log file is still opened, but nothing extra is printed.) I suppose it would be preferable to return immediately from all listener methods if `!active`; this is set pretty early as a `@ClassRule` so there should not be a race condition.